File: jar-auth/src/main/java/com/beancounter/auth/common/TokenService.java
package com.beancounter.auth.common;

import com.beancounter.auth.client.LoginService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken;
import org.springframework.stereotype.Service;

@Service
public class TokenService {

  public static final String BEARER = "Bearer ";

  private LoginService loginService;

  private static boolean isTokenBased(Authentication authentication) {
    return authentication.getClass().isAssignableFrom(JwtAuthenticationToken.class);
  }

  @Autowired(required = false)
  void setLoginService(LoginService loginService) {
    this.loginService = loginService;
  }

  public JwtAuthenticationToken getJwtToken() {
    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
    if (authentication == null) {
      return null;
    }
    if (isTokenBased(authentication)) {
      return (JwtAuthenticationToken) authentication;
    } else {
      return null;
    }
  }

  public String getToken() {
    JwtAuthenticationToken jwt = getJwtToken();
    if (jwt != null) {
      return jwt.getToken().getTokenValue();
    } else {
      if (loginService != null) {
        // M2M token
        return loginService.login();
      }

    }
    return null;
  }

  public String getBearerToken() {
    return getBearerToken(getToken());
  }

  public String getBearerToken(String token) {
    return BEARER + token;
  }

  public String getSubject() {
    JwtAuthenticationToken token = getJwtToken();
    if (token == null) {
      return null;
    }
    return token.getToken().getSubject();
  }
}

File: jar-auth/src/main/java/com/beancounter/auth/common/TokenService.kt
package com.beancounter.auth.common

import com.beancounter.auth.client.LoginService
import com.beancounter.auth.server.AuthConstants
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.Authentication
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.security.oauth2.server.resource.authentication.JwtAuthenticationToken
import org.springframework.stereotype.Service

/**
 * Support for auth tokens.
 */
@Service
class TokenService {
    private var loginService: LoginService? = null

    @Autowired(required = false)
    fun setLoginService(loginService: LoginService?) {
        this.loginService = loginService
    }

    val jwtToken: JwtAuthenticationToken?
        get() {
            val authentication = SecurityContextHolder.getContext().authentication
                ?: return null
            return if (isTokenBased(authentication)) {
                authentication as JwtAuthenticationToken
            } else {
                null
            }
        }

    // M2M token
    val token: String?
        get() {
            val jwt = jwtToken
            if (jwt != null) {
                return jwt.token.tokenValue
            } else {
                if (loginService != null) {
                    // M2M token
                    return loginService!!.login()
                }
            }
            return null
        }
    val bearerToken: String
        get() = getBearerToken(token)

    fun getBearerToken(token: String?): String {
        return BEARER + token
    }

    val subject: String?
        get() {
            val token = jwtToken ?: return null
            return token.token.subject
        }

    val isServiceToken: Boolean
        get() {
            val authentication = SecurityContextHolder.getContext().authentication
                ?: return false
            return authentication.authorities.contains(AuthConstants.AUTH_M2M)
        }

    companion object {
        const val BEARER = "Bearer "
        private fun isTokenBased(authentication: Authentication): Boolean {
            return authentication.javaClass.isAssignableFrom(JwtAuthenticationToken::class.java)
        }
    }
}
